url_matcher/patterns.py

```diff
@@ -0,0 +1,265 @@
+    """
+    Returns the domain of the pattern if any.
+
+    >>> get_pattern_domain("")
+
+    >>> get_pattern_domain("/")
+
+    >>> get_pattern_domain("dom")
+    'dom'
+    >>> get_pattern_domain("DOM")
+    'dom'
+    >>> get_pattern_domain("dom:80")
+    'dom'
+    >>> get_pattern_domain("http://dom:80")
+    'dom'
+    >>> get_pattern_domain("http://dom/a_path")
+    'dom'
+    """
+    parsed = pattern_parse(pattern)
+    if parsed.netloc:
+        return get_domain(f"//{parsed.netloc}").lower()
+    return None
+
+
+def pattern_to_url(pattern: str) -> str:
+    """
+    Required for urlparse to recognize the domain in patterns
+    like example.com/path
+
+    >>> pattern_to_url("example.com/")
+    '//example.com/'
+    >>> pattern_to_url("example.com")
+    '//example.com'
+    >>> pattern_to_url("https://example.com")
+    'https://example.com'
+    >>> pattern_to_url("MySchema4+.-://example.com")
+    'MySchema4+.-://example.com'
+    >>> pattern_to_url("//example.com")
+    '////example.com'
+    """
+    # As defined in https://datatracker.ietf.org/doc/html/rfc3986#section-3.1
+    has_scheme = re.search(r"^([a-zA-Z][a-zA-Z0-9.+-]*:)?//", pattern)
+    if not has_scheme:
+        pattern = f"//{pattern}"
+    elif pattern.startswith("//"):
+        # This is required because urlparse("//example.com").netloc == "//example.com"
+        # but instead we want it to be parsed into the the path. We achieve it by appending
+        # two more slashes
+        pattern = f"//{pattern}"
+    return pattern
+
+
+ParseTuple = namedtuple("ParseTuple", "scheme netloc path query fragment")
+
+
+@lru_cache(30)
+def pattern_parse(pattern: str) -> ParseTuple:
+    """
+    Parses the pattern to a named tuple (scheme, netloc, path, query, fragment)
+    >>> pattern_parse("example.com")
+    ParseTuple(scheme='', netloc='example.com', path='', query='', fragment='')
+    >>> pattern_parse("//example.com/path;this_is_also_path")
+    ParseTuple(scheme='', netloc='', path='//example.com/path;this_is_also_path', query='', fragment='')
+    """
+    pattern = pattern_to_url(pattern)
+    return _urlparse(pattern)
+
+
+def _urlparse(url: str) -> ParseTuple:
+    """
+    Returns a named tuple (scheme, netloc, path, query, fragment)
+    where path and params are joined together into path and
+    some other elements are normalized.
+
+    >>> _urlparse("scheme://example.com/path;params?query=23#fragment")
+    ParseTuple(scheme='scheme', netloc='example.com', path='/path;params', query='query=23', fragment='fragment')
+    >>> _urlparse("http://example.com:80/path")
+    ParseTuple(scheme='http', netloc='example.com', path='/path', query='', fragment='')
+    """
+    scheme, netloc, path, params, query, fragment = urlparse(url)
+    path = _join_path_and_params(path, params)
+    scheme, netloc = normalize_netloc_and_schema(scheme, netloc)
+    return ParseTuple(scheme, netloc, path, query, fragment)
+
+
+def _wildcard_re_escape(text: str):
```

Comment:
Let's add a return annotation for consistency: `-> str`.

Same comment for the function next to this.

Reply:
Thanks! done here: https://github.com/zytedata/url-matcher/pull/2/commits/c71a31bb9aabcdb620860090b954a8e55c024934